migration/promote.sh

```diff
@@ -0,0 +1,118 @@
+  readarray -t directories < <(find "${ROOT}/migration" -maxdepth 1 -type d -name 'v*')
+
+  local directory
+  for directory in "${directories[@]##"${ROOT}/migration/"}"; do
+    ver.parse "${directory}" "${directory}"
+  done
+
+  # the previous - the greatest excluding the current
+  local prev="prev"
+  ver.parse "${prev}" "v0.0"
+
+  for directory in "${directories[@]##"${ROOT}/migration/"}"; do
+    if [[ "${directory}" == "${series}" ]]; then
+      continue
+    elif ver.gt "${directory}" "${prev}"; then
+      prev="${directory}"
+    fi
+  done
+
+  # replace references
+  yq --inplace --front-matter process ".from =\"${prev}\"" "${target_migration}/README.md"
+  yq --inplace --front-matter process ".to =\"${target_version}\"" "${target_migration}/README.md"
+  yq --inplace --front-matter process ".from =\"${target_version}\"" "${ROOT}/migration/main/README.md"
+
+  # update header
+  sed -i "s/# Main upgrade and migration/# ${target_version} upgrade and migration/" "${target_migration}/README.md"
+  # update preamble
+  sed -i "/^<\!-- begin preamble --->$/,/^<\!--- end preamble --->$/c\\
+> [\!important]\\
+> This is the upgrade and migration process for Welkin Apps ${target_version}.\\
+>\\
+> Upgrade is supported from any patch version of the major or minor version stated in the \`from\` field above!" "${target_migration}/README.md"
+  # update changelog
+  sed -i "s#\.\./\.\./changelog#../../changelog/${target_version}.md#" "${target_migration}/README.md"
+  # update fetch
+  sed -i "s/Switch to the main branch and pull the latest changes/Fetch the latest changes and switch to the release tag/" "${target_migration}/README.md"
+  sed -i "s/git switch main/git fetch/" "${target_migration}/README.md"
+  sed -i "s/git pull/git switch -d ${target_version}.z/" "${target_migration}/README.md"
+}
+
+prune() {
+  # find current migration directories
+  local -a directories
+  readarray -t directories < <(find "${ROOT}/migration" -maxdepth 1 -type d -name 'v*')
+
+  # until we have five migrations left
+  while [[ "${#directories[*]}" -gt 5 ]]; do
+    # the last - the least including the current
+    local last="${target_version}"
+    ver.parse "${last}" "${target_version}"
+
+    for directory in "${directories[@]##"${ROOT}/migration/"}"; do
+      if [[ "${directory}" == "${series}" ]]; then
+        continue
+      elif ver.gt "${last}" "${directory}"; then
+        last="${directory}"
+      fi
+    done
+
+    log.info "pruning old migration $(esc.blu "${last}")"
+    rm -r "${ROOT}/migration/${last}"
+
+    readarray -t directories < <(find "${ROOT}/migration" -maxdepth 1 -type d -name 'v*')
+  done
+}
+
+main() {
+  local series="${1:-}"
+
+  [[ -n "${series}" ]] || log.fatal "missing required argument $(esc.ylw "<release-series>"), $(usage)"
+
+  ver.parse next "${series}"
+
+  local target_version="v${ver["next-major"]}.${ver["next-minor"]}"
+  local target_migration="${ROOT}/migration/${target_version}"
+
+  create
+
+  prune
```

Comment:
Not a fan of relying on global variables like this, to me it's always been a footgun as soon as you start refactoring things. Not a blocker though, I'm fine with both.